secret_store/src/listener/service_contract_listener.rs

```diff
@@ -0,0 +1,699 @@
+			self.data.last_retry.store(0, Ordering::Relaxed);
+		}
+	}
+}
+
+impl ClusterSessionsListener<GenerationSession> for ServiceContractListener {
+	fn on_session_removed(&self, session: Arc<GenerationSession>) {
+		// only publish when the session is started by another node
+		// when it is started by this node, it is published from process_service_task
+		if !is_processed_by_this_key_server(&*self.data.key_server_set, &*self.data.self_key_pair, &session.id()) {
+			// by this time sesion must already be completed - either successfully, or not
+			debug_assert!(session.is_finished());
+
+			// ignore result - the only thing that we can do is to log the error
+			let _ = session.wait(Some(Default::default()))
+				.map_err(|e| format!("{}", e))
+				.and_then(|server_key| Self::publish_server_key(&self.data, &session.id(), &server_key))
+				.map(|_| {
+					trace!(target: "secretstore", "{}: completed foreign GenerateServerKey({}) request",
+						self.data.self_key_pair.public(), session.id());
+					()
+				})
+				.map_err(|error| {
+					warn!(target: "secretstore", "{}: failed to process GenerateServerKey({}) request with: {}",
+						self.data.self_key_pair.public(), session.id(), error);
+					error
+				});
+		}
+	}
+}
+
+impl TasksQueue {
+	/// Create new tasks queue.
+	pub fn new() -> Self {
+		TasksQueue {
+			service_event: Condvar::new(),
+			service_tasks: Mutex::new(VecDeque::new()),
+		}
+	}
+
+	/// Shutdown tasks queue.
+	pub fn shutdown(&self) {
+		let mut service_tasks = self.service_tasks.lock();
+		service_tasks.push_front(ServiceTask::Shutdown);
+		self.service_event.notify_all();
+	}
+
+	//// Push new tasks to the queue.
+	pub fn push<I>(&self, tasks: I) where I: Iterator<Item=ServiceTask> {
+		let mut service_tasks = self.service_tasks.lock();
+		service_tasks.extend(tasks);
+		if !service_tasks.is_empty() {
+			self.service_event.notify_all();
+		}
+	}
+
+	/// Wait for new task.
+	pub fn wait(&self) -> ServiceTask {
+		let mut service_tasks = self.service_tasks.lock();
+		if service_tasks.is_empty() {
+			self.service_event.wait(&mut service_tasks);
+		}
+
+		service_tasks.pop_front()
+			.expect("service_event is only fired when there are new tasks or is_shutdown == true; is_shutdown == false; qed")
+	}
+}
+
+/// Returns true when session, related to `server_key_id` must be started on this KeyServer.
+fn is_processed_by_this_key_server(key_server_set: &KeyServerSet, self_key_pair: &NodeKeyPair, server_key_id: &H256) -> bool {
+	let servers = key_server_set.get();
+	let total_servers_count = servers.len();
+	if total_servers_count == 0 {
+		return false;
+	}
+	let this_server_index = match servers.keys().enumerate().find(|&(_, s)| s == self_key_pair.public()) {
+		Some((index, _)) => index,
+		None => return false,
+	};
+
+	let server_key_id_value: U256 = server_key_id.into();
+	let range_interval = U256::max_value() / total_servers_count.into();
+	let range_begin = (range_interval + 1.into()) * this_server_index.into();
```

Comment:
Might panic in case there is only one server (overflow).

Reply:
Thanks, good catch!!!